### DISCUSSION TITLE
*Silences*(spaces) in sound stream on bluetooth speaker

### DISCUSSION THREAD

**Discussion by PowerTcX** [pschatzmann/arduino-audio-tools#490] [Q&A] [answered]
Hi, I got "spaces" in between sound stream, while testing example mp3 file from SD card to Bluetooth speaker, maybe need increase some buffer size ?
Audio file bitrate - 179kbps, got same issue with 128kbps, tried different esp32 boards.
[Recorded audio from a bluetooth speaker](https://drive.google.com/file/d/1XdZ9QigstsOstkjGqe66tu_3YGONsLJ9/view?usp=share_link)

```
/**
 * @file player-sdfat-a2dp.ino
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-sdfat-a2dp/README.md
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"

//const uint8_t SD_CS_PIN = 13;
const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSDFAT source(startFilePath, ext);
A2DPStream out;
MP3DecoderHelix decoder;
AudioPlayer player(source, out, decoder);


void setup() {
  //Serial.begin(115200);
  //AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  SPI.begin(14, 2, 15, 13); //For esp32 TT, for 3.5 comment this/ select TTGO T1
  // setup output - We send the test signal via A2DP - so we conect to the "LEXON MINO L" Bluetooth Speaker
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.name = "JBL GO Essential";
  cfg.auto_reconnect = true;  // if this is use we just quickly connect to the last device ignoring cfg.name
  out.begin(cfg);

  // setup player
  player.setVolume(0.5);
  player.begin();

}

void loop() {
  player.copy();
}
```

**Comment by PowerTcX** [comment]
Here we can see silence in the waveform:
![Screenshot_20221203_120106](https://user-images.githubusercontent.com/71061795/205435384-55cf0037-0b48-4cf4-8a07-cf029ff518bf.jpg)

**Comment by joba-1** [comment]
how did you record this? 1:1 digital bit stream output or with some analog steps involved? It is not complete silence, so without analog steps it can‘t be a buffer issue but a calculation problem.

**Reply by PowerTcX** [reply]
Recorded to my phone, then open with android audio editor application.
So not complete silence = room ambient noise.

**Comment by pschatzmann** [comment]
You did not write how often this happens, so I did some tests myself and for some files I could notice some rate small breaks - maybe 1 per 1 or 2 minutes.

I don't think that this is related to the buffer size: it is aleady very big: 1280 * 50 bytes!

The current implementation is using a RingBuffer which is synchronized with a semaphore. I think the most promising approach would be to replace this with the new [SynchronizedBufferRTOS](https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_synchronized_buffer_r_t_o_s.html) (described [here](https://www.pschatzmann.ch/home/2022/11/25/arduino-audio-tools-multicore-processing/))

Do you want to pick this up ?

**Reply by PowerTcX** [reply]
This happens every 8-12 seconds.

**Reply by pschatzmann** [reply]
If this is happening so often to you, you might have an issue with the read speed ? Did you compare the different SD implementations or play around with the settings in the SFAT  ?
The audio samples need to be provided at a faster fate then 44200 samples per second to ADDP, otherwise it will stotter.

**Reply by PowerTcX** [reply]
`Do you want to pick this up ?`
Later i try this, if i can do this with my weak coding skills.
`If this is happening so often to you, you might have an issue with the read speed ?` 
Maybe, i tried different SD cards, current is "Kingston microSDHC Class 10 UHS-I"
`Did you compare the different SD implementations or play around with the settings in the SFAT ?`
No, I used only one configaration.

**Comment by pschatzmann** [comment, marked as answer]
I made some test sketch with the SynchronizedBufferRTOS using the A2DP API and it is actually worse!
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/sandbox/base-player-a2dp/base-player-a2dp.ino

**Reply by PowerTcX** [reply]
Thank U for helping, for me last example even don’t connect to Bluetooth speaker...

This is  a log output [AudioLogger::Info]
```
[I] AudioStreams.h : 1114 - setVolume: 1.000000
[I] MetaDataID3.h : 544 - virtual void audio_tools::MetaDataID3::begin()
[I] AudioSourceSDFAT.h : 104 - selectStream: 0
[I] AudioSourceSDFAT.h : 121 - -> selectStream: /Oliver Rokk - House of Memories (Slowed).mp3
[I] AudioCopy.h : 68 - buffer_size=1024
[I] AudioPlayer.h : 284 - setVolume(0.100000)
[I] AudioStreams.h : 1114 - setVolume: 0.100000
[I] AudioStreams.h : 1114 - setVolume: 0.100000
[I] AudioPlayer.h : 198 - sample_rate: 44100
[I] AudioPlayer.h : 199 - bits_per_sample: 16
[I] AudioPlayer.h : 200 - channels: 2
[I] AudioStreams.h : 1114 - setVolume: 0.100000
[I] AudioStreams.h : 1114 - setVolume: 0.100000
[I] AudioTypes.h : 55 - sample_rate: 44100
[I] AudioTypes.h : 56 - channels: 2
[I] AudioTypes.h : 57 - bits_per_sample: 16
[I] AudioCopy.h : 133 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 133 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
```

**Reply by pschatzmann** [reply]
If the bluetooth session is not ended properly - e.g. by restarting it can be that the bluetooth speaker hangs and does not allow any connection. When you restart the speaker (and maybe the ESP32 as well) it will work again...
Dn't forget to set the log level to info in the Arduino IDE, so that you see the A2DP messages

**Reply by PowerTcX** [reply]
Yes, i tried reset speaker and esp32 - no result, once i upload first sketch [example](https://github.com/pschatzmann/arduino-audio-tools/discussions/490#discussion-4622269), no more problems with bluetooth connection.

**Reply by PowerTcX** [reply]
Thank U, working great now !

**Comment by pschatzmann** [comment]
I finally managed to have a synchronized buffer implementation with consistent timing, so the [basic example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/base-player-a2dp/base-player-a2dp.ino) is working consistently with perfect qulality for me now.

If you still have issues with this sketch you can try to do some performance measurment on your system:
-[mp3 max decoding speed](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/tests/performance/mp3-Speed/mp3-Speed.ino)
-[max synchronized data receiving speed](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/tests/performance/mp3-SynchronizedNBuffer/mp3-SynchronizedNBuffer.ino)
The values must be well above the 44100 sampling rate!
